llama_index/indices/managed/vectara/retriever.py

```diff
@@ -0,0 +1,150 @@
+"""Vectara index.
+An index that that is built on top of Vectara
+"""
+
+import logging
+from typing import Any, List, Optional, Sequence, Dict
+import json
+
+from llama_index.schema import TextNode
+from llama_index.indices.base_retriever import BaseRetriever
+from llama_index.schema import NodeWithScore
+from llama_index.indices.managed.base import ManagedIndex
+from dataclasses import dataclass
+from llama_index.schema import BaseNode
+
+_logger = logging.getLogger(__name__)
+
+
+@dataclass
+class VectaraQuery:
+    """Vectara query."""
+
+    similarity_top_k: int = 1
+    query_str: Optional[str] = None
+
+    # For Hybrid search: 0 = neural search only, 1 = bm25 only. In between values are a linear interpolation
+    lambda_val: Optional[float] = None
+
+    # define how many sentences before/after the matched sentence to return in the node
+    n_sentences_before: int = 2
+    n_sentences_after: int = 2
+
+    # metadata filters
+    filter: Optional[str] = None
+
+
+@dataclass
+class VectaraQueryResult:
+    """Vectara query result."""
+
+    nodes: Optional[Sequence[BaseNode]] = None
+    similarities: Optional[List[float]] = None
+    ids: Optional[List[str]] = None
+
+
+class VectaraRetriever(BaseRetriever):
+    """Vectara Retriever.
+    Args:
+    index (VectaraIndex): the Vectara Index
+    similarity_top_k (int): number of top k results to return.
+    vectara_kwargs (dict): Additional vectara specific kwargs to pass through to the vectara at query time.
+    """
+
+    def __init__(
+        self,
+        index: ManagedIndex,
+        **kwargs: Any,
+    ) -> None:
+        """Initialize params."""
+        self._index = index
+        self._kwargs: Dict[str, Any] = kwargs.get("vectara_kwargs", {})
+
+    def _get_post_headers(self) -> dict:
+        """Returns headers that should be attached to each post request."""
+        return {
+            "x-api-key": self._index._vectara_api_key,
+            "customer-id": self._index._vectara_customer_id,
+            "Content-Type": "application/json",
+        }
+
+    def _retrieve(
+        self,
+        query: VectaraQuery,
```

Comment:
hmm, slightly torn on this

The `BaseRetriever` class calls `retrieve` with a query bundle (or at least it's typed that way)

Does it make sense to maybe attach the query settings to the retriever, similar to the `VectorStoreRetriever`?

Reply:
yeah I was also torn here but didn't quite know what else to do. Can you please elaborate what exactly you have in mind here?

Reply:
Yea, I think moving the settings to the retriever init makes sense? Similar to the vector store retriever here: https://github.com/jerryjliu/llama_index/blob/b20a495cd71bd50760863566f924375d3aafed05/llama_index/indices/vector_store/retrievers/retriever.py#L55

Reply:
So is the idea that retriever still uses query_bundle? the issue is query_bundle assumes embeddings, which are internal to Vectara, so not that clean. (unless I mis-understand your intent)
Or would VectaraRetriever store all the query parameters internally, and then receive just a query_str?

Reply:
The query bundle actually just assumes text. You can optionally attach embeddings fro the vector index, but we can ignore that here.

But yea, the retriever would store all the parameters, and received a query bundle (which is really just a string 99% of the time, we can ignore the embeddings if a user tries to attach them)

Reply:
Updated. Can you pls check if this fits your suggestion?